docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,123 @@
+    { }
+}
+```
+
+Troublesome APIs:
+- [`Span<T>`](https://docs.microsoft.com/dotnet/api/system.span-1)
+    - `public Span(T[]? array);`
+    - `public Span(T[]? array, int start, int length);`
+    - `public T[] ToArray();`
+    - `public static implicit operator Span<T>(ArraySegment<T> segment);`
+    - `public static implicit operator Span<T>(T[]? array);`
+- [`ReadOnlySpan<T>`](https://docs.microsoft.com/dotnet/api/system.readonlyspan-1)
+    - `public ReadOnlySpan(T[]? array);`
+    - `public ReadOnlySpan(T[]? array, int start, int length);`
+    - `public T[] ToArray();`
+    - `public static implicit operator ReadOnlySpan<T>(ArraySegment<T> segment);`
+    - `public static implicit operator ReadOnlySpan<T>(T[]? array);`
+
+A new `GenericParameterAttributes` value will be defined which also represents metadata defined in the `CorGenericParamAttr` enumeration. Space is provided between the existing constraints group to permit constraint growth while keeping the bit-mask contiguous.
+
+```diff
+namespace System.Reflection
+{
+    [Flags]
+    public enum GenericParameterAttributes
+    {
++        AcceptByRefLike = 0x0100
+    }
+}
+```
+
+```diff
+typedef enum CorGenericParamAttr
+{
++   gpAcceptByRefLike = 0x0100 // type argument can be ByRefLike
+} CorGenericParamAttr;
+```
+
+The expansion of metadata will impact at least the following:
+
+- ILDasm/ILAsm &ndash; https://github.com/dotnet/runtime
+- Cecil &ndash; https://github.com/jbevain/cecil
+- IL Trimmer &ndash; https://github.com/dotnet/linker
+- F# &ndash; https://github.com/fsharp/fsharp
+- C++/CLI &ndash; The MSVC team
+
+## Semantic Proposal
+
+An API that is a JIT-time intrinsic will be needed to determine if a parameter is ByRefLike. This API would represent a check to occur at JIT time to avoid taking paths that would be invalid for some values of `T`. The existing `Type.IsByRefLike` property will be made an intrinsic (e.g., `typeof(T).IsByRefLike`).
+
+For dispatch to object implemented methods and to default interface methods, the behavior shall be that an `InvalidProgramException` should be thrown. The JIT will insert the following IL at code-gen time.
+
+```
+newobj instance void System.InvalidProgramException::.ctor()
+throw
+```
+
+The `Reflection.Emit` API will be updated to respect ByRefLike support in Generics. The current Reflection API requires boxing of all types which makes usage of ByRefLike types impossible. Until the Reflection API can fully support ByRefLike types, this feature must be blocked when using Reflection. For example, API calls such as `MakeGenericType` / `MakeGenericMethod` will be invalid when `T` is ByRefLike.
+
+## Special IL Sequences
+
+The following are IL sequences involving the `box` instruction. They are used in scenario optimizations and shall continue to be valid, even with ByRefLike types, in cases where the result can be computed at JIT time and elided safely. They will be added to the ECMA-335 addendum.
```

Comment:
```suggestion
The following are IL sequences involving the `box` instruction. They are used for common C# language constructs and shall continue to be valid, even with ByRefLike types, in cases where the result can be computed at JIT time and elided safely. They will be added to the ECMA-335 addendum.
```

Reply:
> in cases where the result can be computed at JIT time

What are those cases exactly? There are varying levels of optimizations and assumptions that the JIT can apply to these. The exact cases that are required to be have recognized by the JIT have to be enumerated as part of the spec. These cases will be required for functional correctness, they won't be an optional optimization anymore.

Reply:
I think you probably want to say that the JIT is always required to recognize these sequences and never perform the box operation for them. 

Whether the JIT can compute the result as a constant or whether it calls a helper function to compute the result is JIT implementation detail.